### DISCUSSION TITLE
Best practices prediction

### DISCUSSION THREAD

**Discussion by jyrr** [awslabs/gluonts#948] [Q&A]
I'm having trouble understanding why we need to supplement a target for predictions? If a model is trained, say up till date X and I want to predict from date X+1, my predictions without supplying target seem random at best. However, if I set target to e.g. a day worth of data before X+1 the predictions are usable.

I'm wondering what the best practice is? The data I'm using has 15-minute frequency, I'm using a prediction length of 96 (one day). I'm using DeepAREstimator. If I supply the day before as data (for prediction) the predictions seem usable, however if I supply 2 or 3 days the predictions tend to get notably worse (by a factor 2 or 3).

**Comment by AaronSpieler** [comment]
We call the time series of the target variable we want to predict `target`. 
Usually we split our datasets so that the test set is `prediction_length` longer so that we can evaluate our Predictors on previously unseen data.
If you wish to predict into the future, you will still have to provide `target` however, only for time points that lie in the past.

**Comment by jyrr** [comment]
> We call the time series of the target variable we want to predict `target`.
> Usually we split our datasets so that the test set is `prediction_length` longer so that we can evaluate our Predictors on previously unseen data.
> If you wish to predict into the future, you will still have to provide `target` however, only for time points that lie in the past.

Hey, thanks for the reply. I understand all of what you're saying, apologies if my first message was a bit unclear.
I'm specifically asking about a trained model. I understand that, at least for the DeepAR model, to predict on new data, we'll have to supply it with data in `gluon-ts` format (containing `start` and `target`). It will then return `prediction_length` worth of predictions.

```
predict_ds = ListDataset(
    [{'start': args.start, 'target': data}],
    freq=args.freq
)

predictions = list(predictor.predict(predict_ds))
mean_predictions = np.mean(predictions[0].samples, axis=0)
```

It's just that I've never before had to give an already trained model data in order to predict, this was always only the case for training the model itself, thus I'm trying to wrap my head around this.

Let's say I have a (fictional) data set with `frequency='H'` and `prediction_length=24` from 1/1/2020 up to and including 31/12/2020. I trained my model on this data set (ignoring test set for simplicity) and now I want to predict 1/1/2021. What do I do here? Supply the full year as data, or last month, week, day, hour? Are there any best practices or would this be completely model dependent?

Adding to that, if I were to predict 2/1/2021 (so another `prediction_length`) worth, how do I go about this? Call the models `train` method with the 1/1/2021 data, or as per above?

Just trying to make sense of it all, thank you!

**Comment by AaronSpieler** [comment]
From what you wrote it seems to me that you might not have explicitly set `context_length`. 
In DeepAR we set it like this internally:
```
self.context_length = (
            context_length if context_length is not None else prediction_length
        )
```
So equal to the prediction length if you have not set it explicitly.

Now, `context_length` specifies how much the model looks back in the time series to generate a prediction of length `prediction_length`. This is also how the model is trained on the training data set (we automatically split the training time series into a `prediction length` and `context_length` part).

Then, if you only provide 1 to 3 time steps in the past for the trained model (i.e. the Predictor), you give it less data then it had usually during training to generate its predictions. The difference in performance between supplying 1 or 3 days might just be random, since both are significantly less than for example the 24 you talked about above.

In general I would expect all models to take some input to generate predictions, since providing no input (in this case no target) would leave the model no choice but to predict the average of all the training samples, independent of input (in a way the same as having no input).

You could try setting `context_lenght` to 0 for training, which would be like not using data to predict, however, I'm not sure whether we support this.

**Comment by AaronSpieler** [comment]
If this answered your question @jyrr , could you please close the issue?

**Comment by jyrr** [comment]
I'm quite sure it does not predict the average of all training samples without providing input, the predictions returned were in [1e-11,1e-10] range where as the actual data is in [0,1000].

Anyhow, if I understand correctly I should give the Predictor `context_length` of data to get `prediction_length` worth of predictions. However what I don't understand is, if I give the Predictor `> context_length` (for example 2, 3 or 4x) of data I notice the predictions become significantly worse (evaluating on a holdout set). Is this normal?

Another example; (sorry, as a novice they really help me learn), I have a year of data, `freq=D`, `prediction_length=7=context_length`.  I want to predict for the full next year. What would be the best way to go about this? From what I understand;

- Train the model on the full year of data (with the proper params)
- For the prediction of the first week of the new year, use the last week of the past year as input for the predictor (which was also in the train data)
- For the prediction of each next week, use last weeks (real) data.

I would assume, considering volatility, it's best to periodically 'update' the model, eg train, per X weeks. Is that a correct assumption?

If so, and if giving the Predictor `> context_length` of data (and perhaps `< context_length`, haven't checked) should indeed result in significantly worse performance, then I understand.

**Reply by pratikgehlott** [reply]
@jyrr, could you see this last comment, and answer it, please.
